### PR TITLE
Adds support for FreeBSD

### LINKED ISSUE #607: Add compiler support for FreeBSD
I tried to compile wazero for FreeBSD, but got an error.

```
../../../../pkg/mod/github.com/tetratelabs/wazero@v0.0.0-20220529032218-029a79476b4f/internal/engine/compiler/mmap.go:74:16: undefined: syscall.Mprotect
```

Looks like `Mprotect` is not implemented in `syscall`.
https://github.com/golang/go/blob/aac1d3a1b12a290805ca35ff268738fb334b1ca4/src/syscall/zsyscall_freebsd_amd64.go

It is implemented in `golang.org/x/sys/unix`. I confirmed I could compile it after replacing `syscall` with `golang.org/x/sys/unix`.
https://github.com/golang/sys/blob/bc2c85ada10aa9b6aa9607e9ac9ad0761b95cf1d/unix/zsyscall_freebsd_amd64.go#L296-L308

If wazero needs to keep zero dependency (otherwise, need to rename it to `waone`?), can't we even use `golang.org/x`? Do you want to re-implement syscalls for other platforms in wazero?

### PR DESCRIPTION
**Pull request by codefromthecrypt**
FreeBSD was disabled due to lack of testing. This works around the
compilation problems. Note: We can't currently test arm64 automatically!

Notes:

* GitHub Actions doesn’t support FreeBSD, and may never.
* We could use Travis to run FreeBSD, but it would split our CI config.
* Using Vagrant directly is easier to debug than vmactions/freebsd-vm.
* GitHub Actions only supports virtualization on MacOS.
* GitHub Actions removed vagrant from the image starting with macos-11.
* Since VirtualBox doesn't work on arm64, freebsd/arm64 is untestable.

Fixes #607

### CODE REVIEW

**Review by mathetake — APPROVED** [review]
Awesome (glad it works without changing anything except mmap)

**Comment by codefromthecrypt** [comment]
heh let's see.. I accidentally missed the build flags to actually use it with compiler 🤞

**Comment by codefromthecrypt** [comment]
I can't currently reproduce the failure locally with vagrant even if deleting the VM and doing it again.

```bash
$ vagrant destroy
$ go version
go version go1.18.3 darwin/amd64
$ GOVERSION=$(go env GOVERSION) GOARCH=$(go env GOARCH) vagrant up --provision
--snip--
$ vagrant ssh -c "cd wazero && make spectest.v1"
ok  	github.com/tetratelabs/wazero/internal/integration_test/spectest/v1	1.854s
Connection to 127.0.0.1 closed.
```

**Comment by codefromthecrypt** [comment]
can't try anymore because vagrant rate limits (via HTTP 429)

will park this meanwile

**Comment by mathetake** [comment]
cool passed!

File: internal/engine/compiler/impl_amd64.go [hunk 1]
```diff
@@ -1229,7 +1229,7 @@ func (c *amd64Compiler) compileCtz(o *wazeroir.OperationCtz) error {
 		return err
 	}
 
-	if runtime.GOOS != "darwin" {
+	if runtime.GOOS != "darwin" && runtime.GOOS != "freebsd" {
```
Comment (mathetake) [inline internal/engine/compiler/impl_amd64.go:1232]: maybe this is not need and i guess only change on ^ `compileClz ` was necessary but not sure

**Pull request merged**